File: mcp-server/src/test/kotlin/com/riwonace/mcp/tools/SqlGuardTest.kt
package com.riwonace.mcp.tools

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class SqlGuardTest {

    @Test
    fun `단순 SELECT는 LIMIT이 보강되어 통과한다`() {
        val sql = SqlGuard.sanitize("SELECT name, salary FROM employees WHERE dept = '플랫폼팀'")
        assertTrue(sql.endsWith("LIMIT 50"))
    }

    @Test
    fun `이미 LIMIT이 있으면 그대로 통과한다`() {
        val sql = SqlGuard.sanitize("select * from products order by price desc limit 5;")
        assertEquals("select * from products order by price desc limit 5", sql)
    }

    @Test
    fun `WITH 절 SELECT도 허용한다`() {
        val sql = SqlGuard.sanitize("WITH t AS (SELECT dept FROM employees) SELECT dept FROM t")
        assertTrue(sql.startsWith("WITH"))
    }

    @Test
    fun `DML과 DDL은 거부한다`() {
        assertThrows<IllegalArgumentException> { SqlGuard.sanitize("DELETE FROM employees") }
        assertThrows<IllegalArgumentException> { SqlGuard.sanitize("DROP TABLE products") }
        assertThrows<IllegalArgumentException> { SqlGuard.sanitize("UPDATE products SET price = 0") }
    }

    @Test
    fun `다중 문장과 주석 인젝션은 거부한다`() {
        assertThrows<IllegalArgumentException> {
            SqlGuard.sanitize("SELECT 1; DROP TABLE employees")
        }
        assertThrows<IllegalArgumentException> {
            SqlGuard.sanitize("SELECT * FROM employees -- hidden")
        }
    }

    @Test
    fun `SELECT로 위장한 위험 함수는 거부한다`() {
        assertThrows<IllegalArgumentException> { SqlGuard.sanitize("SELECT pg_sleep(10)") }
    }

    @Test
    fun `대소문자 섞은 우회 시도 차단`() {
        assertThrows<IllegalArgumentException> { SqlGuard.sanitize("sElEcT DeLeTe FROM users") }
        assertThrows<IllegalArgumentException> { SqlGuard.sanitize("SELECT InSeRt FROM users") }
    }

    @Test
    fun `위험 함수 우회 시도 차단`() {
        assertThrows<IllegalArgumentException> { SqlGuard.sanitize("SELECT PG_SLEEP(5)") }
        assertThrows<IllegalArgumentException> { SqlGuard.sanitize("SELECT pg_read_file('/etc/passwd')") }
        assertThrows<IllegalArgumentException> { SqlGuard.sanitize("SELECT lo_import('/etc/passwd')") }
    }

    @Test
    fun `서브쿼리 내 위험 함수 차단`() {
        assertThrows<IllegalArgumentException> {
            SqlGuard.sanitize("SELECT * FROM (SELECT pg_sleep(1)) t")
        }
    }

    @Test
    fun `INTO 키워드 차단`() {
        assertThrows<IllegalArgumentException> {
            SqlGuard.sanitize("SELECT * INTO new_table FROM employees")
        }
    }
}

File: mcp-server-go/internal/tools/sqlguard_test.go
package tools

import (
	"strings"
	"testing"
)

// Test names mirror SqlGuardTest.kt cases 1:1 for functional-equivalence review.

func TestSanitizeSQL_SimpleSelectGetsLimitAppended(t *testing.T) {
	sql, err := SanitizeSQL("SELECT name, salary FROM employees WHERE dept = '플랫폼팀'")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !strings.HasSuffix(sql, "LIMIT 50") {
		t.Fatalf("expected LIMIT 50 suffix, got %q", sql)
	}
}

func TestSanitizeSQL_ExistingLimitPassesThrough(t *testing.T) {
	sql, err := SanitizeSQL("select * from products order by price desc limit 5;")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	want := "select * from products order by price desc limit 5"
	if sql != want {
		t.Fatalf("got %q, want %q", sql, want)
	}
}

func TestSanitizeSQL_WithClauseAllowed(t *testing.T) {
	sql, err := SanitizeSQL("WITH t AS (SELECT dept FROM employees) SELECT dept FROM t")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !strings.HasPrefix(sql, "WITH") {
		t.Fatalf("expected WITH prefix, got %q", sql)
	}
}

func TestSanitizeSQL_DmlAndDdlRejected(t *testing.T) {
	for _, sql := range []string{
		"DELETE FROM employees",
		"DROP TABLE products",
		"UPDATE products SET price = 0",
	} {
		if _, err := SanitizeSQL(sql); err == nil {
			t.Fatalf("expected rejection for %q", sql)
		}
	}
}

func TestSanitizeSQL_MultiStatementAndCommentInjectionRejected(t *testing.T) {
	if _, err := SanitizeSQL("SELECT 1; DROP TABLE employees"); err == nil {
		t.Fatal("expected rejection for multi-statement SQL")
	}
	if _, err := SanitizeSQL("SELECT * FROM employees -- hidden"); err == nil {
		t.Fatal("expected rejection for comment injection")
	}
}

func TestSanitizeSQL_DangerousFunctionDisguisedAsSelectRejected(t *testing.T) {
	if _, err := SanitizeSQL("SELECT pg_sleep(10)"); err == nil {
		t.Fatal("expected rejection for pg_sleep")
	}
}

func TestSanitizeSQL_MixedCaseBypassRejected(t *testing.T) {
	if _, err := SanitizeSQL("sElEcT DeLeTe FROM users"); err == nil {
		t.Fatal("expected rejection for mixed-case DELETE bypass")
	}
	if _, err := SanitizeSQL("SELECT InSeRt FROM users"); err == nil {
		t.Fatal("expected rejection for mixed-case INSERT bypass")
	}
}

func TestSanitizeSQL_DangerousFunctionBypassRejected(t *testing.T) {
	if _, err := SanitizeSQL("SELECT PG_SLEEP(5)"); err == nil {
		t.Fatal("expected rejection for PG_SLEEP")
	}
	if _, err := SanitizeSQL("SELECT pg_read_file('/etc/passwd')"); err == nil {
		t.Fatal("expected rejection for pg_read_file")
	}
	if _, err := SanitizeSQL("SELECT lo_import('/etc/passwd')"); err == nil {
		t.Fatal("expected rejection for lo_import")
	}
}

func TestSanitizeSQL_DangerousFunctionInSubqueryRejected(t *testing.T) {
	if _, err := SanitizeSQL("SELECT * FROM (SELECT pg_sleep(1)) t"); err == nil {
		t.Fatal("expected rejection for pg_sleep in subquery")
	}
}

func TestSanitizeSQL_IntoKeywordRejected(t *testing.T) {
	if _, err := SanitizeSQL("SELECT * INTO new_table FROM employees"); err == nil {
		t.Fatal("expected rejection for INTO keyword")
	}
}
